File: java/src/test/java/TestCase.java
import com.saucelabs.saucebindings.SauceOptions;
import com.saucelabs.saucebindings.SauceSession;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.rules.TestName;
import org.junit.rules.TestWatcher;
import org.junit.runner.Description;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;

public class TestCase {

    protected WebDriver driver;

    @Rule
    public TestName testName = new TestName() {
        public String getMethodName() {
            return String.format("%s", super.getMethodName());
        }
    };

    @Rule
    public SauceTestWatcher sauceTestWatcher = new SauceTestWatcher();

    @Before
    public void setUp() {
        SauceOptions sauceOptions = new SauceOptions();
        sauceOptions.setCapability("browserName", "chrome");
        sauceOptions.setCapability("browserVersion", "latest");
        sauceOptions.setName(testName.getMethodName());
        sauceOptions.setBuild("Wayfair-Java-Example");

        SauceSession session = new SauceSession(sauceOptions);
        driver = session.start();
    }

    public class SauceTestWatcher extends TestWatcher {
        @Override
        protected void succeeded(Description description) {
            if (driver != null) {
                ((JavascriptExecutor) driver).executeScript("sauce:job-result=passed");
                driver.quit();
            }
            super.succeeded(description);
        }

        @Override
        protected void failed(Throwable e, Description description) {
            if (driver != null) {
                ((JavascriptExecutor) driver).executeScript("sauce:job-result=failed");
                driver.quit();
            }
            super.failed(e, description);
        }
    }

}

File: kotlin/src/test/kotlin/TestCase.kt
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.TestInfo
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.api.extension.TestWatcher
import org.openqa.selenium.JavascriptExecutor
import org.openqa.selenium.MutableCapabilities
import org.openqa.selenium.WebDriver
import org.openqa.selenium.chrome.ChromeOptions
import org.openqa.selenium.remote.RemoteWebDriver
import java.net.URL


open class TestCase {

    protected lateinit var driver: WebDriver
    protected lateinit var js: JavascriptExecutor

    class SauceTestWatcher : TestWatcher {

        public override fun testSuccessful(context: ExtensionContext?) {
            js.execute()
        }
    }

    @BeforeEach
    fun setUp(info: TestInfo) {
        val sauceOptions = MutableCapabilities()
        val browserOptions = ChromeOptions()

        browserOptions.setCapability("platformName", "Windows 10")
        browserOptions.setCapability("browserVersion", "latest")
        sauceOptions.setCapability("name", info.displayName.toString())
        sauceOptions.setCapability("build", "Kotlin-Wayfair-Example")
        browserOptions.setCapability("sauce:options", sauceOptions)

        val username = System.getenv("SAUCE_USERNAME")
        val accessKey = System.getenv("SAUCE_ACCESS_KEY")
        val url = "https://$username:$accessKey@ondemand.us-west-1.saucelabs.com/wd/hub/"

        val remoteUrl = URL(url)

        driver = RemoteWebDriver(remoteUrl, browserOptions)
    }

    @AfterEach
    fun tearDown() {
        js = driver as JavascriptExecutor
        driver.quit()
    }
}